server/src/crm/crm.service.ts

```diff
@@ -66,340 +116,46 @@ export class CrmService {
-    //  get token
-    const JWToken = await this.ADAL.acquireToken();
-    const options = {
-      url: `${this.host}${query}`,
-      headers: {
-        'Accept-Encoding': 'gzip, deflate',
-        'Content-Type': 'application/json; charset=utf-8',
-        Authorization: `Bearer ${JWToken}`,
-        'OData-MaxVersion': '4.0',
-        'OData-Version': '4.0',
-        Accept: 'application/json',
-        Prefer: 'odata.include-annotations="*"',
-        ...headers
-      },
-      encoding: null,
-    };
-
-    return new Promise((resolve, reject) => {
-      Request.get(options, (error, response, body) => {
-        const encoding = response.headers['content-encoding'];
-
-        if (!error && response.statusCode === 200) {
-          // If response is gzip, unzip first
 
-          const parseResponse = jsonText => {
-            const json_string = jsonText.toString('utf-8');
+      return acc;
+    }, {});
+    const oDataReq = new ORequest(this.config.get('CRM_HOST'), {});
 
-            var result = JSON.parse(json_string, this._dateReviver);
-            if (result["@odata.context"].indexOf("/$entity") >= 0) {
-              // retrieve single
-              result = this._fixLongODataAnnotations(result);
-            }
-            else if (result.value) {
-              // retrieve multiple
-              var array = [];
-              for (var i = 0; i < result.value.length; i++) {
-                array.push(this._fixLongODataAnnotations(result.value[i]));
-              }
-              result.value = array;
-            }
-            resolve(result);
-          };
+    oDataReq.applyQuery(truthyKeyedObject);
 
-          if (encoding && encoding.indexOf('gzip') >= 0) {
-            zlib.gunzip(body, (err, dezipped) => {
-              parseResponse(dezipped);
-            });
-          }
-          else {
-            parseResponse(body);
-          }
-        } else {
-          const parseError = jsonText => {
-            // Bug: sometimes CRM returns 'object reference' error
-            // Fix: if we retry error will not show again
-            const json_string = jsonText.toString('utf-8');
-            const result = JSON.parse(json_string, this._dateReviver);
-            const err = this._parseErrorMessage(result);
+    const { url: { search } } = oDataReq;
 
-            if (err == "Object reference not set to an instance of an object.") {
-              this._get(query, maxPageSize, options)
-                .then(
-                  resolve, reject
-                );
-            } else {
-              reject(err);
-            }
-          };
-
-          if (encoding && encoding.indexOf('gzip') >= 0) {
-            zlib.gunzip(body, (err, dezipped) => {
-              parseError(dezipped);
-            });
-          } else {
-            parseError(body);
-          }
-        }
-      });
-    });
-  }
-
-  async _create(query, data, headers): Promise<any> {
```

Comment:
👍🏾  Looks like we don't use _create in either Search or Portal API, so deleting is ok

Reply:
will re-add in follow up but it's simpler for me to not create methods that don't have any direct usage in the app itself